File: Objects/MatchScore.vb
Imports System.Math

<Serializable()> _
Public Class MatchScore
	Inherits BusinessBase

	Private Shared log As ILog = LogManager.GetLogger(GetType(MatchScore).ToString())

	''' <summary>
	''' MatchSet represents a single set from a match score.
	''' </summary>
	''' <remarks>
	''' MatchSet can be used to represent a match score in one of two ways: objectively and subjectively.
	''' 
	''' When used objectively, WGames and LGames represent the number of games won by the winner and loser, 
	''' respectively.  In this case, WGames is interpreted as "winner games won" (games won in this set by
	''' the winner of the match), and LGames is "loser games won".
	''' 
	''' When used subjectively, WGames and LGames represent the number of games won and lost
	''' by a specific player.   In this case, WGames and LGames are interpreted as "games won and lost by the
	''' player the player whose record is being retrieved".
	''' 
	''' MatchSet does not know which way it is being used, as it does not contain any information about the
	''' players.  It is up to the parent class to know which way it is using MatchSet.
	''' 
	''' Note that WGames and LGames are named with just 'W' and 'L' in order
	''' </remarks>
	<Serializable()> _
	Public Class MatchSet
		Inherits BusinessBase
		Private _WGames As Byte
		Private _LGames As Byte
		Private _WTiebreak As Byte
        Private _LTiebreak As Byte
        Private _SetNumber As Integer

        Friend Sub New(ByVal setNumber As Integer)
            _SetNumber = setNumber
        End Sub

        'if the tiebreak was played, the winner of the tiebreak must also be the winner of the set
        'todo: this rule is not true if the loser defaulted the match during the tiebreak
        'in this case the game score would be tied, and the tiebreak score could be anything
        Private Function IsScoreValid() As Boolean
            If TiebreakPlayed AndAlso _WGames <> _LGames Then
                Return Sign(_WGames.CompareTo(_LGames)) = Sign(_WTiebreak.CompareTo(_LTiebreak))
            End If
            Return True
        End Function

        Private Sub ValidateScore()
            'BrokenRules.Assert(propertyName & "_Score", "Score must be valid", propertyName, Not IsScoreValid())
            BrokenRules.Assert("Set" & _SetNumber & "_Score", String.Format("Set {0} winner must have also won the tiebreak", _SetNumber), "LTiebreak", Not IsScoreValid())
        End Sub

        Public Property WGames() As Byte
            Get
                Return _WGames
            End Get
            Set(ByVal value As Byte)
                If _WGames = value Then Return
                _WGames = value
                MarkDirty()
                ValidateProperty("WGames", _WGames)
                ValidateScore()
            End Set
        End Property
        Public Property LGames() As Byte
            Get
                Return _LGames
            End Get
            Set(ByVal value As Byte)
                If _LGames = value Then Return
                _LGames = value
                MarkDirty()
                ValidateProperty("LGames", _LGames)
                ValidateScore()
            End Set
        End Property
        Public Property WTiebreak() As Byte
            Get
                Return _WTiebreak
            End Get
            Set(ByVal value As Byte)
                If _WTiebreak = value Then Return
                _WTiebreak = value
                MarkDirty()
                ValidateProperty("WTiebreak", _WTiebreak)
                ValidateScore()
            End Set
        End Property
        Public Property LTiebreak() As Byte
            Get
                Return _LTiebreak
            End Get
            Set(ByVal value As Byte)
                If _LTiebreak = value Then Return
                _LTiebreak = value
                MarkDirty()
                ValidateProperty("LTiebreak", _LTiebreak)
                ValidateScore()
            End Set
        End Property

        Public ReadOnly Property SetPlayed() As Boolean
            Get
                Return WGames > 0 Or LGames > 0
            End Get
        End Property
        Public ReadOnly Property TiebreakPlayed() As Boolean
            Get
                Return WTiebreak > 0 Or LTiebreak > 0
            End Get
        End Property
        Public Overrides Function ToString() As String
            Return ToString(False)
        End Function
        Public Overloads Function ToString(ByVal IncludeTiebreaks As Boolean) As String
            If Not SetPlayed Then
                Return ""
            ElseIf TiebreakPlayed And IncludeTiebreaks Then
                Return String.Format("{0}-{1} ({2}-{3})", WGames, LGames, WTiebreak, LTiebreak)
            Else
                Return String.Format("{0}-{1}", WGames, LGames)
            End If
        End Function

        Friend Sub Fetch(ByVal dr As SafeDataReader)
            WGames = dr.GetByte
            LGames = dr.GetByte
            WTiebreak = dr.GetByte
            LTiebreak = dr.GetByte
            MarkOld()
        End Sub

        'Public Overrides ReadOnly Property IsValid() As Boolean
        '	Get
        '		If Not MyBase.IsValid Then Return False
        '              'If TiebreakPlayed AndAlso _WGames <> _LGames Then
        '              '	Return _WGames.CompareTo(_LGames) <> _WTiebreak.CompareTo(_LTiebreak)
        '              'End If
        '              'Return True
        '	End Get
        'End Property
    End Class

#Region "Private Data"
	Private _Sets(4) As MatchSet
	Private _LoserDefaulted As Boolean
#End Region

#Region "Business Properties and Methods"
	''' <summary>
	''' Returns the specified MatchSet object for the match (0-based).
	''' </summary>
	''' <param name="index">0-based index of the set</param>
	''' <value></value>
	''' <returns></returns>
	''' <remarks></remarks>
	Public ReadOnly Property Sets(ByVal index As Integer) As MatchSet
		Get
			Return _Sets(index)
		End Get
	End Property

	''' <summary>
	''' Returns 0-based array of MatchSets for the match.
	''' This version is for C# compatibility (the Sets property is not
	''' available in C#).
	''' </summary>
	''' <value></value>
	''' <returns></returns>
	''' <remarks></remarks>
	Public ReadOnly Property MatchSets() As MatchSet()
		Get
			Return _Sets
		End Get
	End Property

	Public ReadOnly Property ShowSet(ByVal index As Integer) As Boolean
		Get
			Return index = 0 OrElse _Sets(index - 1).SetPlayed
		End Get
	End Property

	Public Property LoserDefaulted() As Boolean
		Get
			Return _LoserDefaulted
		End Get
		Set(ByVal value As Boolean)
			If _LoserDefaulted = value Then Return
			_LoserDefaulted = value
			MarkDirty()
			'ValidateProperty("LoserDefaulted", _LoserDefaulted)
		End Set
	End Property

	Public ReadOnly Property WSets() As Int32
		Get
			Dim n As Int32 = 0
			For i As Integer = 0 To 4
				If _Sets(i).WGames > _Sets(i).LGames Then n += 1
			Next
			Return n
		End Get
	End Property

	Public ReadOnly Property LSets() As Int32
		Get
			Dim n As Int32 = 0
			For i As Integer = 0 To 4
				If _Sets(i).LGames > _Sets(i).WGames Then n += 1
			Next
			Return n
		End Get
	End Property

#End Region

#Region " IsValid and IsDirty "
	Public Overrides ReadOnly Property IsValid() As Boolean
		Get
            If Not MyBase.IsValid Then Return False
			For i As Integer = 0 To 4
				If Not _Sets(i).IsValid Then Return False
			Next

			'if loser defaulted, no extra validation is needed
			If _LoserDefaulted Then Return True

			'otherwise check validity of match score
			'winner must have won more sets that loser
			If WSets <= LSets Then Return False

			Return True
		End Get
	End Property

	Public Overrides ReadOnly Property IsDirty() As Boolean
		Get
			If MyBase.IsDirty Then Return True
			For i As Integer = 0 To 4
				If _Sets(i).IsDirty Then Return True
			Next
		End Get
	End Property

    'Sub Set_IsDirtyChanged(ByVal sender As Object, ByVal e As EventArgs)
    '	Me.OnIsDirtyChanged()
    'End Sub
#End Region

#Region "System.Object Overrides"
	Public Overrides Function ToString() As String
		Return ToString(False)
	End Function

	Public Overloads Function ToString(ByVal IncludeTiebreaks As Boolean) As String
		Dim s As New System.Text.StringBuilder
		For i As Integer = 0 To 4
			s.Append(_Sets(i).ToString(IncludeTiebreaks))
			s.Append(" ")
		Next
		Return s.ToString.Trim()
	End Function

	Public Overloads Function Equals(ByVal A As MatchScore) As Boolean
		Return MyBase.Equals(A)
	End Function

	Public Overrides Function GetHashCode() As Integer
		Return MyBase.GetHashCode()
	End Function
#End Region

#Region "Constructors"
	Friend Sub New()
		'Prevent direction creation
		MarkAsChild()
		For i As Int16 = 0 To 4
            _Sets(i) = New MatchSet(i + 1)
            AddHandler _Sets(i).IsDirtyChanged, AddressOf Child_IsDirtyChanged
		Next
	End Sub
#End Region

#Region "Data Access"
	Friend Sub Fetch(ByVal dr As SafeDataReader)
		For i As Integer = 0 To 4
			_Sets(i).Fetch(dr)
		Next
		_LoserDefaulted = dr.GetBoolean
		MarkOld()
		LoadRules()
	End Sub
#End Region

End Class

File: TennisObjects/MatchScore.cs
using System;
using System.Collections.Generic;
using System.Text;
using Csla;
using log4net;
using Bahns.Data;
using Csla.Validation;


namespace TennisObjects
{
	[Serializable()]
	public class MatchScore : BusinessBase<MatchScore>
	{

		private static ILog log = LogManager.GetLogger(typeof(MatchScore).ToString());

		/// <summary> 
		/// MatchSet represents a single set from a match score. 
		/// </summary> 
		/// <remarks> 
		/// MatchSet can be used to represent a match score in one of two ways: objectively and subjectively. 
		/// 
		/// When used objectively, WGames and LGames represent the number of games won by the winner and loser, 
		/// respectively. In this case, WGames is interpreted as "winner games won" (games won in this set by 
		/// the winner of the match), and LGames is "loser games won". 
		/// 
		/// When used subjectively, WGames and LGames represent the number of games won and lost 
		/// by a specific player. In this case, WGames and LGames are interpreted as "games won and lost by the 
		/// player the player whose record is being retrieved". 
		/// 
		/// MatchSet does not know which way it is being used, as it does not contain any information about the 
		/// players. It is up to the parent class to know which way it is using MatchSet. 
		/// 
		/// Note that WGames and LGames are named with just 'W' and 'L' in order 
		/// </remarks> 
		[Serializable()]
		public class MatchSet : BusinessBase<MatchSet>
		{
			private byte _WGames;
			private byte _LGames;
			private byte _WTiebreak;
			private byte _LTiebreak;
			private int _SetNumber;
			private MatchScore _parent;

			internal void MatchScore_Saved(object sender, Csla.Core.SavedEventArgs e)
			{
				//MatchScore parent = (MatchScore)e.NewObject;
				MarkOld();
			}

			internal MatchSet(int setNumber, MatchScore parent)
			{
				_SetNumber = setNumber;
				_parent = parent;
			}

			//if the tiebreak was played, the winner of the tiebreak must also be the winner of the set 
			//todo: this rule is not true if the loser defaulted the match during the tiebreak 
			//in this case the game score would be tied, and the tiebreak score could be anything 
			private bool IsScoreValid()
			{
				if (TiebreakPlayed && _WGames != _LGames)
				{
					return Math.Sign(_WGames.CompareTo(_LGames)) == Math.Sign(_WTiebreak.CompareTo(_LTiebreak));
				}
				return true;
			}

			public bool ValidSetScore(object target, RuleArgs e)
			{
				if (!_parent.LoserDefaulted && TiebreakPlayed)
				{
					e.Description = string.Format("The set {0} winner must have won the tiebreak.", _SetNumber);
					return Math.Sign(_WGames.CompareTo(_LGames)) == Math.Sign(_WTiebreak.CompareTo(_LTiebreak));
				}

				//commented out because the parent is responsible for checking the overall match score
				//i.e. the parent calls AddInstanceBusinessRule to register ValidMatchScore, so we don't
				//     need to call it here
				//return _parent.ValidMatchScore(_parent, e);
				return true;
			}

			protected override void AddBusinessRules()
			{
				//ValidationRules.AddRule(IsScoreValid, "WGames");
				//ValidationRules.AddRule(CommonRules.StringRequired, new RuleArgs("FirstName", "First name"));
				//ValidationRules.AddRule(CommonRules.StringRequired, new RuleArgs("LastName", "Last name"));

				ValidationRules.AddDependantProperty("WGames", "LGames", true);
				ValidationRules.AddDependantProperty("WGames", "WTiebreak", true);
				ValidationRules.AddDependantProperty("WGames", "LTiebreak", true);
			}

			protected override void AddInstanceBusinessRules()
			{
				ValidationRules.AddInstanceRule(ValidSetScore, "WGames");
				//ValidationRules.AddInstanceRule(ValidSetScore, "LGames");
				//ValidationRules.AddInstanceRule(ValidSetScore, "WTiebreak");
				//ValidationRules.AddInstanceRule(ValidSetScore, "LTiebreak");
			}

			//private bool SetWinnerIsTiebreakWinner(object target, Csla.Validation.RuleArgs e)
			//{
			//    //if loser defaulted, no extra validation is needed 
			//    if (_LoserDefaulted)
			//        return true;

			//    //otherwise check validity of match score 
			//    //winner must have won more sets than loser 
			//    if (WSets <= LSets)
			//        return false;


			//    return true;
			//}


			private void ValidateScore()
			{
				//BrokenRules.Assert(propertyName & "_Score", "Score must be valid", propertyName, Not IsScoreValid()) 
				//BrokenRules.Assert("Set" + _SetNumber + "_Score", string.Format("Set {0} winner must have also won the tiebreak", _SetNumber), "LTiebreak", !IsScoreValid());
			}

			public byte WGames
			{
				get { return _WGames; }
				set { SetProperty(ref _WGames, value); }
			}
			public byte LGames
			{
				get { return _LGames; }
				set { SetProperty(ref _LGames, value); }
			}
			public byte WTiebreak
			{
				get { return _WTiebreak; }
				set { SetProperty(ref _WTiebreak, value); }
			}
			public byte LTiebreak
			{
				get { return _LTiebreak; }
				set { SetProperty(ref _LTiebreak, value); }
			}

			public bool SetPlayed
			{
				get { return WGames > 0 | LGames > 0; }
			}
			public bool TiebreakPlayed
			{
				get { return WTiebreak > 0 | LTiebreak > 0; }
			}
			public override string ToString()
			{
				return ToString(false);
			}
			public string ToString(bool IncludeTiebreaks)
			{
				if (!SetPlayed)
				{
					return "";
				}
				else if (TiebreakPlayed & IncludeTiebreaks)
				{
					return string.Format("{0}-{1} ({2}-{3})", WGames, LGames, WTiebreak, LTiebreak);
				}
				else
				{
					return string.Format("{0}-{1}", WGames, LGames);
				}
			}

			internal void Fetch(SafeDataReader dr)
			{
				WGames = dr.GetByte();
				LGames = dr.GetByte();
				WTiebreak = dr.GetByte();
				LTiebreak = dr.GetByte();
				MarkOld();
			}

			protected override object GetIdValue()
			{
				throw new Exception("The method or operation is not implemented.");
			}
		}

		#region Private Data
		private MatchSet[] _Sets = new MatchSet[5];
		private bool _LoserDefaulted;
		#endregion

		#region Business Properties and Methods
		/// <summary> 
		/// Returns the specified MatchSet object for the match (0-based). 
		/// </summary> 
		/// <param name="index">0-based index of the set</param> 
		/// <value></value> 
		/// <returns></returns> 
		/// <remarks></remarks> 
		/*public MatchSet Sets
		{
			get { return _Sets[index]; }
		}*/

		public MatchSet[] Sets
		{
			get { return _Sets; }
		}

		/// <summary> 
		/// Returns 0-based array of MatchSets for the match. 
		/// This version is for C# compatibility (the Sets property is not 
		/// available in C#). 
		/// </summary> 
		/// <value></value> 
		/// <returns></returns> 
		/// <remarks></remarks> 
		public MatchSet[] MatchSets
		{
			get { return _Sets; }
		}

		public bool ShowSet(int index)
		{
			return index == 0 || _Sets[index - 1].SetPlayed;
		}

		public bool LoserDefaulted
		{
			get { return _LoserDefaulted; }
			set
			{
				if (_LoserDefaulted == value)
					return;
				_LoserDefaulted = value;
				//MarkDirty();
				//ValidateProperty("LoserDefaulted", _LoserDefaulted) 
				PropertyHasChanged();
			}
		}

		public Int32 WSets
		{
			get
			{
				Int32 n = 0;
				for (int i = 0; i <= 4; i++)
				{
					if (_Sets[i].WGames > _Sets[i].LGames)
						n += 1;
				}
				return n;
			}
		}

		public Int32 LSets
		{
			get
			{
				Int32 n = 0;
				for (int i = 0; i <= 4; i++)
				{
					if (_Sets[i].LGames > _Sets[i].WGames)
						n += 1;
				}
				return n;
			}
		}

		#endregion

		#region Validation
		protected override void AddBusinessRules()
		{
			//ValidationRules.AddRule(IsScoreValid, "WGames");
			//ValidationRules.AddRule(CommonRules.StringRequired, new RuleArgs("FirstName", "First name"));
			//ValidationRules.AddRule(CommonRules.StringRequired, new RuleArgs("LastName", "Last name"));

		}

		protected override void AddInstanceBusinessRules()
		{
			ValidationRules.AddInstanceRule(ValidMatchScore, "LoserDefaulted");
		}

		private bool ValidMatchScore(object target, Csla.Validation.RuleArgs e)
		{
			//if loser didn't default, winner must have won more sets than loser 
			e.Description = "Winner must have won more sets than loser.";
			return _LoserDefaulted || WSets > LSets;
		}

		#endregion

		#region IsValid and IsDirty
		public override bool IsValid
		{
			get
			{
				if (!base.IsValid)
					return false;

				for (int i = 0; i <= 4; i++)
				{
					if (!_Sets[i].IsValid)
						return false;
				}

				return true;
			}
		}

		public override bool IsDirty
		{
			get
			{
				if (base.IsDirty)
					return true;
				for (int i = 0; i <= 4; i++)
				{
					if (_Sets[i].IsDirty)
						return true;
				}
				return false;
			}
		}

		//Sub Set_IsDirtyChanged(ByVal sender As Object, ByVal e As EventArgs) 
		// Me.OnIsDirtyChanged() 
		//End Sub 

		public override List<BrokenRule> GetAllBrokenRules()
		{
			List<BrokenRule> rules = base.GetAllBrokenRules();
			for (int i = 0; i <= 4; i++)
			{
				foreach (BrokenRule rule in _Sets[i].GetAllBrokenRules())
					rules.Add(rule);
			}
			return rules;
		}

		#endregion

		#region System.Object Overrides
		public override string ToString()
		{
			return ToString(false);
		}

		public string ToString(bool IncludeTiebreaks)
		{
			System.Text.StringBuilder s = new System.Text.StringBuilder();
			for (int i = 0; i <= 4; i++)
			{
				s.Append(_Sets[i].ToString(IncludeTiebreaks));
				s.Append(" ");
			}
			return s.ToString().Trim();
		}

		protected override object GetIdValue()
		{
			throw new Exception("The method or operation is not implemented.");
		}
		#endregion

		#region Constructors
		internal MatchScore()
		{
			//Prevent direct creation 
			MarkAsChild();
			for (Int16 i = 0; i <= 4; i++)
			{
				_Sets[i] = new MatchSet(i + 1, this);
				_Sets[i].PropertyChanged += Child_PropertyChangedGeneric;
				_Sets[i].PropertyChanged += new System.ComponentModel.PropertyChangedEventHandler(MatchScore_PropertyChanged);
				Saved += new EventHandler<Csla.Core.SavedEventArgs>(_Sets[i].MatchScore_Saved);
			}
		}

		void MatchScore_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
		{
			//e.PropertyName;
			ValidationRules.CheckRules("LoserDefaulted");
		}

		internal void Match_Saved(object sender, Csla.Core.SavedEventArgs e)
		{
			MarkOld();
		}

		#endregion

		#region Data Access
		internal void Fetch(SafeDataReader dr)
		{
			for (int i = 0; i <= 4; i++)
			{
				_Sets[i].Fetch(dr);
			}
			_LoserDefaulted = dr.GetBoolean();
			MarkOld();
		}
		#endregion


	}

}
